src/lib.rs

```diff
@@ -67,3 +88,77 @@ pub use errors::{Error, ErrorKind, Result};
 const APP_INFO: app_dirs::AppInfo = app_dirs::AppInfo {name: "Tectonic", author: "TectonicProject"};
 
 const FORMAT_SERIAL: u32 = 28; // keep synchronized with tectonic/xetex-constants.h!!
+
+
+/// Compile LaTeX text to a PDF.
+///
+/// This function is an all-in-one interface to the main Tectonic workflow. Given
+/// a string representing a LaTeX input file, it will compile it to a PDF and return
+/// a byte vector corresponding to the resulting file:
+///
+/// ```
+/// let latex = r#"
+/// \documentclass{article}
+/// \begin{document}
+/// Hello, world!
+/// \end{document}
+/// "#;
+///
+/// # tectonic::test_util::activate_test_mode_augmented(env!("CARGO_MANIFEST_DIR"));
+/// let pdf_data: Vec<u8> = tectonic::latex_to_pdf(latex).expect("processing failed");
+/// println!("Output PDF size is {} bytes", pdf_data.len());
+/// ```
+///
+/// The compilation uses the default bundle, the location of which is embedded
+/// in the crate or potentially specified in the user’s configuration file.
+/// The current working directory will be searched for any `\\input` files.
+/// Messages aimed at the user are suppressed, but (in the default
+/// configuration) network I/O may occur to pull down needed resource files.
+/// No outputs are written to disk; all supporting files besides the PDF
+/// document are discarded. The XeTeX engine is run multiple times if needed
+/// to get the output file to converge.
+///
+/// For more sophisticated uses, use the [`driver`] module, which provides a
+/// high-level interface for driving the typesetting engines with much more
+/// control over their behavior.
+pub fn latex_to_pdf<T: AsRef<str>>(latex: T) -> Result<Vec<u8>> {
```

Comment:
Is tectonic thread safe now? If I remember correctly, there used to be unsynchronized static globals in the C core, and the access to the bundle is not synchronized. It seems that this function lets the user run multiple instances of the engine in parallel.

Reply:
Good point. We are still not thread safe, and probably won't be for a long time: the TeX engine just has a *lot* of shared global state in the C code. It will take a *long* time to put that all into some kind of non-global struct.

At the moment I can think of two solutions:

1. Bite the bullet and add locking to the engines.
2. Skirt the issue and just mention in this function's documentation that it is not thread-safe in the way you describe.

I lean towards option 2. Thoughts?

Reply:
I would strongly prefer option 1. Rust selling point is "threads without data races", enforced by the type system. A quick and dirty solution is to have a Mutex that has to be acquired at the beginning of `latex_to_pdf`(similar to the test infrastructure currently), and add a note in the documentation that at most one instance of the engine can be running at a time and so the function might block until the other instances are finished.

Reply:
That's definitely a concern. One way to communicate this would be introducing a singleton struct, say `Tectonic`, that would have methods `latex_to_pdf` and `processing_session_builder`. The user would have to first create the struct using `Tectonic::new()`, which would return an error if `Tectonic` were already in use.

Reply:
I think that approach is a little less desirable because one day, hopefully, we will be thread-safe, and then the singleton `Tectonic` will be unnecessary and a bit inconvenient.

I'll investigate adding locking within the crate and see how it works out.

Reply:
Yeah, that was pretty easy to implement. I tend to think this is the right way to go.